### DISCUSSION TITLE
addWeekDay method to DateTime

### DISCUSSION THREAD

**Discussion by FMorschel** [jogboms/time.dart#58] [Ideas]
Maybe creating an `addWorkDay` method to `DateTime`? I suggest something like the following (edited, found a broken case on the first example - when starting on weekends it would add another day-, and added the `subtractWorkDays` method):

```dart
extension AddWorkDays on DateTime {

  bool get isWeekend =>
      (weekday == DateTime.saturday) || (weekday == DateTime.sunday);

  bool get isWorkDay => !isWeekend;

  bool get _notFridayNorSaturday =>
      (weekday != DateTime.friday) && (weekday != DateTime.saturday);

  DateTime addWorkDays(int days) {
    final dayToAdd = (days.isNegative ? -1 : 1).days;
    if (isWorkDay && (days == 0)) return this;
    if (days.isNegative) return subtractWorkDays(-days);
    if (_notFridayNorSaturday) {
      return add(dayToAdd).addWorkDays(days - 1);
    } else {
      return add(dayToAdd).addWorkDays(days);
    }
  }

  bool get _notMondayNorSunday =>
      (weekday != DateTime.monday) && (weekday != DateTime.sunday);

  DateTime subtractWorkDays(int days) {
    if (isWorkDay && (days == 0)) return this;
    if (days.isNegative) return addWorkDays(-days);
    final dayToAdd = (days.isNegative ? -1 : 1).days;
    if (_notMondayNorSunday) {
      return subtract(dayToAdd).subtractWorkDays(days - 1);
    } else {
      return subtract(dayToAdd).subtractWorkDays(days);
    }
  }
}
```

I'm not sure how useful this would be and about the duration days being added or if it would be better to create another `DateTime` with the day field increased/decreased by one every time.

**Comment by jogboms** [comment]
This is nice. I particularly like the `isWeekend` and `isWorkDay`.

There is no problem AFAIK with using `add` and `subtract` as they would also create a new `DateTime` every time.

**Reply by FMorschel** [reply]
### I have some doubts:
- About the parameters being `int`, do you think that's the best approach? Have any suggestions on this?
- Do you think this would be useful?
- Do you think this should be added to the package even though it's not as useful? (For answering this consider my thoughts stated next)

### Thoughts: 
I was wondering if maybe we could create a new class or an easy-to-use method so that *instead of this*, we could say which days inside the week the package user wants to skip when adding/subtracting days.

**Reply by jogboms** [reply]
Well,
- I think keeping `int`s is good for now and if the need for fractions comes up we could easily use a `num` without introducing a breaking change.
- `isWeekend` & `isWorkDay` I believe I would use. Still, I'm not sure of the value of `addWorkDays` & `subtractWorkDays` for the average joe as it is peculiar to a kind of application. I would not like to bloat the API unless requested by maybe a couple of people.

I believe the second bullet should address the third doubt and the final thoughts on bloating the API layer

**Comment by FMorschel** [comment]
As per [your comment](https://github.com/jogboms/time.dart/discussions/55#discussioncomment-3454677) on #55, I believe this should be closed. Do you think the `isWeekend` and `isWorkDay` should be implemented on a PR?

**Reply by jogboms** [reply]
I think you can, it sounds good to me @FMorschel
